docs/arch/adr-004-python-model-arch.md

```diff
@@ -0,0 +1,60 @@
+In the same idea of dbt is not your query engine, we don't want dbt to be your python runtime. Instead, we want dbt to focus on being the place to express transformation logic. So python model will be following the existing pattern of the SQL model(parse and compile user written logic and submit it to your computation engine).
+
+#### What are the guardrails dbt would enforce for the python model
+We want dbt to focus on transformation logic, so we opt for setting up some tools and guardrails for the python model to focus on doing data transformation.
+1. A `dbt` object would have functions including `dbt.ref`, `dbt.source` function to reference other models and sources in the dbt project, the return of the function will be a dataframe of referenced resources. 
+1. Code in the python model node should include a model function that takes a `dbt` object as an argument, do the data transformation logic inside, and return a dataframe in the end. We think folks should load their data into dataframes using the `dbt.ref`, `dbt.source` provided over raw data references. We also think logic to write dataframe to database objects should live in materialization logic instead of transformation code.
+1. That `dbt` object should also have an attribute called `dbt.config` to allow users to define configurations of the current python model like materialization logic, a specific version of python libraries, etc. This `dbt.config` object should also provide a clear access function for variables defined in project YAML. This way user can access arbitrary configuration at runtime.
+
+#### What logic should live in dbt-core
+
+1. parsing of python file in dbt-core to get the `ref`, `source`, and `config` information. This information is used to place the python model in the correct place in project DAG and generate the correct python code sent to compute engine. 
+1. `language` as a new top-level node property.
+1. python template code that is not cloud provider-specific, this includes implementation for `dbt.ref`, `dbt.source`. We would use ast parser to parse out all of the `dbt.ref`, `dbt.source` inside python during parsing time, and generate what database resources those points to during compilation time. This should allow user to copy-paste the "compiled" code, and run it themselves against the data warehouse — just like with SQL models. A example of definition for `dbt.ref` could look like this
+    ```python
+    def ref(*args):
+    refs = {"my_sql_model": "DBT_TEST.DBT_SOMESCHEMA.my_sql_model"}
+    key = ".".join(args)
+    return load_df_function(refs[key])
+    ```
+
+1. functional tests for the python model, these tests are expected to be inherited in the adapter code to make sure intended functions are met.
+1. Generalizing the names of properties (`sql`, `raw_sql`, `compiled_sql`) for a future where it's not all SQL.
+
+This means adapters will need to implement/modify computing engine specific functions including but not limited to
```

Comment:
So we are saying these items listed all will include specific logic per data warehouse and that is why they need to live in Adapters code?

Reply:
Yeah maybe be a little more explicit that the logic in Core is universal and the Adapter logic is very warehouse specific. The `This means adapters...` sentence sounds like whatever we didn't put in Core we are just putting in Adapters instead of there being concrete reasons